File: codes/matlab/ch04/lagrange_basis.m
%% lagrange_basis.m
%
% Visualizes Lagrange basis polynomials for equispaced vs Chebyshev nodes.
%
% The Lagrange basis polynomials L_k(x) satisfy L_k(x_j) = delta_{kj}.
% For equispaced nodes, they develop large oscillations near boundaries.
% For Chebyshev nodes, they remain well-behaved.
%
% Author: Dr. Denys Dutykh
%         Mathematics Department
%         Khalifa University of Science and Technology
%         Abu Dhabi, UAE
%
% Part of the book "Computational Études: A Spectral Approach"

clear; close all; clc;

%% Configuration
N = 10;
N_FINE = 1000;

% Book color scheme
NAVY = [20, 45, 110] / 255;
SKY = [120, 150, 210] / 255;
CORAL = [231, 76, 60] / 255;
TEAL = [22, 160, 133] / 255;
PURPLE = [142, 68, 173] / 255;
ORANGE = [230, 126, 34] / 255;

% Output path
script_dir = fileparts(mfilename('fullpath'));
output_dir = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch04', 'matlab');
output_file = fullfile(output_dir, 'lagrange_basis.pdf');

%% Generate nodes
x_equi = linspace(-1, 1, N+1)';
j = 0:N;
x_cheb = cos(j * pi / N)';

% Fine grid
x_fine = linspace(-1, 1, N_FINE)';

% Indices to plot
indices = [1, 2, N/2+1, N, N+1];  % 1-indexed
colors = {CORAL, ORANGE, TEAL, SKY, PURPLE};

%% Create figure
fig = figure('Units', 'inches', 'Position', [1, 1, 11, 4.5]);

% Left panel: Equispaced basis functions
subplot(1, 2, 1);
hold on; box on;

for i = 1:length(indices)
    k = indices(i);
    color = colors{i};
    L_k = lagrange_basis_k(x_equi, k, x_fine);
    plot(x_fine, L_k, 'Color', color, 'LineWidth', 1.2, ...
         'DisplayName', sprintf('$L_{%d}(x)$', k-1));
end

% Mark nodes
plot(x_equi, zeros(size(x_equi)), 'o', 'Color', NAVY, 'MarkerSize', 5, ...
     'MarkerFaceColor', NAVY, 'HandleVisibility', 'off');

yline(0, '-', 'Color', [0.5, 0.5, 0.5], 'LineWidth', 0.5, 'HandleVisibility', 'off');
yline(1, '--', 'Color', [0.5, 0.5, 0.5], 'LineWidth', 0.5, 'Alpha', 0.5, 'HandleVisibility', 'off');

xlabel('$x$', 'Interpreter', 'latex');
ylabel('$L_k(x)$', 'Interpreter', 'latex');
title(sprintf('Equispaced Nodes ($N = %d$)', N), 'FontSize', 11);
xlim([-1, 1]);
ylim([-2.5, 2.5]);
legend('Location', 'northwest', 'Interpreter', 'latex', 'NumColumns', 2);

text(-0.5, -2.2, 'Large oscillations near boundaries', 'FontSize', 9, 'Color', [0.5, 0.5, 0.5]);

set(gca, 'FontSize', 10, 'TickLabelInterpreter', 'latex');
grid on;
set(gca, 'GridAlpha', 0.2);

% Right panel: Chebyshev basis functions
subplot(1, 2, 2);
hold on; box on;

for i = 1:length(indices)
    k = indices(i);
    color = colors{i};
    L_k = lagrange_basis_k(x_cheb, k, x_fine);
    plot(x_fine, L_k, 'Color', color, 'LineWidth', 1.2, ...
         'DisplayName', sprintf('$L_{%d}(x)$', k-1));
end

% Mark nodes
plot(x_cheb, zeros(size(x_cheb)), 'o', 'Color', NAVY, 'MarkerSize', 5, ...
     'MarkerFaceColor', NAVY, 'HandleVisibility', 'off');

yline(0, '-', 'Color', [0.5, 0.5, 0.5], 'LineWidth', 0.5, 'HandleVisibility', 'off');
yline(1, '--', 'Color', [0.5, 0.5, 0.5], 'LineWidth', 0.5, 'Alpha', 0.5, 'HandleVisibility', 'off');

xlabel('$x$', 'Interpreter', 'latex');
ylabel('$L_k(x)$', 'Interpreter', 'latex');
title(sprintf('Chebyshev Nodes ($N = %d$)', N), 'FontSize', 11);
xlim([-1, 1]);
ylim([-2.5, 2.5]);
legend('Location', 'northwest', 'Interpreter', 'latex', 'NumColumns', 2);

text(0.3, 1.5, 'Bounded oscillations', 'FontSize', 9, 'Color', [0.5, 0.5, 0.5]);

set(gca, 'FontSize', 10, 'TickLabelInterpreter', 'latex');
grid on;
set(gca, 'GridAlpha', 0.2);

%% Save figure
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

exportgraphics(fig, output_file, 'ContentType', 'vector');
fprintf('Figure saved to: %s\n', output_file);

png_file = strrep(output_file, '.pdf', '.png');
exportgraphics(fig, png_file, 'Resolution', 300);

%% Print maximum values
fprintf('\nMaximum |L_k(x)| for N = %d:\n', N);
fprintf('%s\n', repmat('-', 1, 50));
fprintf('%4s %15s %15s\n', 'k', 'Equispaced', 'Chebyshev');
fprintf('%s\n', repmat('-', 1, 50));
for k = 1:N+1
    L_equi = lagrange_basis_k(x_equi, k, x_fine);
    L_cheb = lagrange_basis_k(x_cheb, k, x_fine);
    fprintf('%4d %15.4f %15.4f\n', k-1, max(abs(L_equi)), max(abs(L_cheb)));
end
fprintf('%s\n', repmat('-', 1, 50));

% Compute Lebesgue constants
lebesgue_equi = zeros(size(x_fine));
lebesgue_cheb = zeros(size(x_fine));
for k = 1:N+1
    lebesgue_equi = lebesgue_equi + abs(lagrange_basis_k(x_equi, k, x_fine));
    lebesgue_cheb = lebesgue_cheb + abs(lagrange_basis_k(x_cheb, k, x_fine));
end
fprintf('\nLebesgue constants:\n');
fprintf('  Equispaced: Lambda_%d = %.2f\n', N, max(lebesgue_equi));
fprintf('  Chebyshev:  Lambda_%d = %.2f\n', N, max(lebesgue_cheb));
fprintf('  Ratio: %.1fx\n', max(lebesgue_equi)/max(lebesgue_cheb));

%% Lagrange basis function
function L_k = lagrange_basis_k(x_nodes, k, x_eval)
    n = length(x_nodes);
    L_k = ones(size(x_eval));

    for j = 1:n
        if j ~= k
            L_k = L_k .* (x_eval - x_nodes(j)) / (x_nodes(k) - x_nodes(j));
        end
    end
end

File: codes/julia/ch04/lagrange_basis.jl
# lagrange_basis.jl
#
# Visualizes Lagrange basis polynomials for equispaced vs Chebyshev nodes.
#
# The Lagrange basis polynomials are:
#     L_k(x) = prod_{j != k} (x - x_j) / (x_k - x_j)
#
# These polynomials satisfy L_k(x_j) = delta_{kj} (Kronecker delta), making them
# the "cardinal functions" for interpolation: the interpolant is simply
#     p_N(x) = sum_{k=0}^{N} f_k L_k(x)
#
# For equispaced nodes, the L_k(x) develop large oscillations near the boundaries,
# contributing to the Runge phenomenon. For Chebyshev nodes, the basis functions
# remain well-behaved across the entire interval.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LaTeXStrings

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const N = 10       # Number of intervals (N+1 nodes)
const N_FINE = 1000

# Book color scheme
const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#8E44AD"
const ORANGE = colorant"#E67E22"

# Output path
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch04", "julia")

# -----------------------------------------------------------------------------
# Node generation
# -----------------------------------------------------------------------------
equispaced_nodes(N) = collect(range(-1, 1, length=N+1))
chebyshev_nodes(N) = [cos(j * π / N) for j in 0:N]

# -----------------------------------------------------------------------------
# Lagrange basis computation
# -----------------------------------------------------------------------------
"""
    lagrange_basis(x_nodes, k, x_eval)

Compute the k-th Lagrange basis polynomial at points `x_eval`.
    L_k(x) = prod_{j != k} (x - x_j) / (x_k - x_j)

Note: `k` uses 1-based indexing.
"""
function lagrange_basis(x_nodes, k, x_eval)
    n = length(x_nodes)
    L_k = ones(length(x_eval))

    for j in 1:n
        if j != k
            L_k .*= (x_eval .- x_nodes[j]) ./ (x_nodes[k] - x_nodes[j])
        end
    end

    return L_k
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Fine grid for plotting
    x_fine = collect(range(-1, 1, length=N_FINE))

    # Generate nodes
    x_equi = equispaced_nodes(N)
    x_cheb = chebyshev_nodes(N)

    # Create figure with two panels
    fig = Figure(size = (1100, 450))

    # Indices of basis functions to plot (1-based: corresponds to k=0,1,N/2,N-1,N)
    indices_to_plot = [1, 2, N÷2+1, N, N+1]
    colors = [CORAL, ORANGE, TEAL, SKY, PURPLE]
    # Labels use 0-based k for display
    labels_k = [0, 1, N÷2, N-1, N]

    # Left panel: Equispaced basis functions
    ax1 = Axis(fig[1, 1],
               xlabel = L"x", ylabel = L"L_k(x)",
               title = "Equispaced Nodes (N = $N)",
               limits = ((-1, 1), (-2.5, 2.5)))

    for (k, col, k_label) in zip(indices_to_plot, colors, labels_k)
        L_k = lagrange_basis(x_equi, k, x_fine)
        lines!(ax1, x_fine, L_k, color = col, linewidth = 1.2,
               label = latexstring("L_{$k_label}(x)"))
    end

    # Mark nodes
    scatter!(ax1, x_equi, zeros(N+1), color = NAVY, markersize = 8)

    hlines!(ax1, 0, color = colorant"#888888", linewidth = 0.5)
    hlines!(ax1, 1, color = colorant"#888888", linewidth = 0.5, linestyle = :dash)

    # Annotation
    text!(ax1, -0.5, -2.2, text = "Large oscillations\nnear boundaries",
          fontsize = 9, color = :gray)

    axislegend(ax1, position = :lt, framevisible = false, labelsize = 9, nbanks = 2)

    # Right panel: Chebyshev basis functions
    ax2 = Axis(fig[1, 2],
               xlabel = L"x", ylabel = L"L_k(x)",
               title = "Chebyshev Nodes (N = $N)",
               limits = ((-1, 1), (-2.5, 2.5)))

    for (k, col, k_label) in zip(indices_to_plot, colors, labels_k)
        L_k = lagrange_basis(x_cheb, k, x_fine)
        lines!(ax2, x_fine, L_k, color = col, linewidth = 1.2,
               label = latexstring("L_{$k_label}(x)"))
    end

    # Mark nodes
    scatter!(ax2, x_cheb, zeros(N+1), color = NAVY, markersize = 8)

    hlines!(ax2, 0, color = colorant"#888888", linewidth = 0.5)
    hlines!(ax2, 1, color = colorant"#888888", linewidth = 0.5, linestyle = :dash)

    # Annotation
    text!(ax2, 0.3, 1.5, text = "Bounded oscillations",
          fontsize = 9, color = :gray)

    axislegend(ax2, position = :lt, framevisible = false, labelsize = 9, nbanks = 2)

    # Save figure
    mkpath(OUTPUT_DIR)
    save(joinpath(OUTPUT_DIR, "lagrange_basis.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "lagrange_basis.png"), fig, px_per_unit = 4)

    println("Figure saved to: ", joinpath(OUTPUT_DIR, "lagrange_basis.pdf"))

    # Print maximum values of basis functions
    println("\nMaximum |L_k(x)| for N = $N:")
    println("-" ^ 50)
    println("$(rpad("k", 4)) $(lpad("Equispaced", 15)) $(lpad("Chebyshev", 15))")
    println("-" ^ 50)
    for k in 1:N+1
        L_equi = lagrange_basis(x_equi, k, x_fine)
        L_cheb = lagrange_basis(x_cheb, k, x_fine)
        println("$(rpad(k-1, 4)) $(lpad(round(maximum(abs.(L_equi)), digits=4), 15)) $(lpad(round(maximum(abs.(L_cheb)), digits=4), 15))")
    end
    println("-" ^ 50)

    # Compute Lebesgue functions
    lebesgue_equi = sum(abs.(lagrange_basis(x_equi, k, x_fine)) for k in 1:N+1)
    lebesgue_cheb = sum(abs.(lagrange_basis(x_cheb, k, x_fine)) for k in 1:N+1)

    println("\nLebesgue constants (Lambda_N = max sum|L_k|):")
    println("  Equispaced: Lambda_$N = $(round(maximum(lebesgue_equi), digits=2))")
    println("  Chebyshev:  Lambda_$N = $(round(maximum(lebesgue_cheb), digits=2))")
    println("  Ratio: $(round(maximum(lebesgue_equi) / maximum(lebesgue_cheb), digits=1))x")

    return fig
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
